### DISCUSSION TITLE
Search by ID?

### DISCUSSION THREAD

**Discussion by stephenleo** [milvus-io/milvus#4743] [Q&A and General discussion] [answered]
I have a collection of 100Million `| product_id | vector |` in a collection.
I would like to find the top_k for any product within the above 100Million products.
Is there a simple way to do this?

For example:
My collection contains
```
| product_id_1 | vector_1 |
| product_id_2 | vector_2 |
| ...          | ...      |
| product_id_N | vector_N |
```
I want to find the top_k `product_ids` for `product_id_1`.

My current approach is use `get_entity_by_id(product_id_1)` to find out `vector_1` and then use `search(vector_1)` to get the top_k. Wondering if there is a simpler one step process?

I found similar issue as well: https://github.com/milvus-io/milvus/issues/3203

Thanks

**Comment by yhmo** [comment, marked as answer]
To support search by id, we need to add an api search_by_id(). In fact, this api is a simple combination of two steps: get_entity_by_id() + search(). Currently milvus has no such api search_by_id().
From our opinion, two apis take a little more effort than one api, so we recommend user takes two steps to do this.

**Reply by stephenleo** [reply]
Got it. thanks.

**Reply by DachuanZhao** [reply]
Two steps will cost more time on network IO ...

**Reply by yhmo** [reply]
> Two steps will cost more time on network IO ...

The v1.x has no search_by_id() interface, so user have to workaround. In v2.0, we will provide search_by_id() .

**Reply by buptzouy** [reply]
Does the java. sdk have similar method of get_entity_by_id()?

**Reply by yhmo** [reply]
> Does the java. sdk have similar method of get_entity_by_id()?

In v1.x, this api is implemented in java sdk:
```
/**
   * Gets vectors data by id array
   *
   * @param collectionName collection to get vectors from
   * @param partitionTag partition to get vectors from
   * @param ids a <code>List</code> of vector ids
   * @return <code>GetEntityByIDResponse</code>
   * @see GetEntityByIDResponse
   * @see Response
   */
  GetEntityByIDResponse getEntityByID(String collectionName, String partitionTag, List<Long> ids);
```

In v2.0, this api is replaced by query() interface.

**Reply by buptzouy** [reply]
> > Does the java. sdk have similar method of get_entity_by_id()?
> 
> In v1.x, this api is implemented in java sdk:
> 
> ```
> /**
>    * Gets vectors data by id array
>    *
>    * @param collectionName collection to get vectors from
>    * @param partitionTag partition to get vectors from
>    * @param ids a <code>List</code> of vector ids
>    * @return <code>GetEntityByIDResponse</code>
>    * @see GetEntityByIDResponse
>    * @see Response
>    */
>   GetEntityByIDResponse getEntityByID(String collectionName, String partitionTag, List<Long> ids);
> ```
> 
> In v2.0, this api is replaced by query() interface.

Thanks for your help,  I have another issue. When I invoke insert, the api return success immediately, but when I invoke query, the java api hangs and never return. Below is my query example code:, it hangs at the code line "R<QueryResults> response = milvusClient.query(test);" 
```  
private R<QueryResults> query() {
        List<Long> queryIDs = new ArrayList<>();
        queryIDs.add(433269551713230010L);
        queryIDs.add(433269551713230011L);
        String expr = "id" + " in " + queryIDs.toString();
        List<String> fields = Arrays.asList( "feature");
        QueryParam test = QueryParam.newBuilder()
                .withCollectionName(COLLECTION_NAME)
                .withExpr(expr)
                .withOutFields(fields)
                .build();
        R<QueryResults> response = milvusClient.query(test);
        handleResponseStatus(response);
        QueryResultsWrapper wrapper = new QueryResultsWrapper(response.getData());
        System.out.println(ID_FIELD + ":" + wrapper.getFieldWrapper(ID_FIELD).getFieldData().toString());
        System.out.println(AGE_FIELD + ":" + wrapper.getFieldWrapper(AGE_FIELD).getFieldData().toString());
        System.out.println("Query row count: " + wrapper.getFieldWrapper(ID_FIELD).getRowCount());
        return response;
    }
```

**Reply by yhmo** [reply]
I can't reproduce.
Maybe the java sdk version is not compatible to milvus version. What is the mivlus version and java sdk version?
